### PR TITLE
improvement: Add atomic update on TrieMap

### PR DESCRIPTION
**Pull request by jkciesluk**
connected to https://github.com/scalameta/metals/issues/5072

### CODE REVIEW

File: mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  private val concurrentMap = new ConcurrentHashMap[K, V]
+
+  def get(key: K): Option[V] = trieMap.get(key)
+
+  def contains(key: K): Boolean = trieMap.contains(key)
+
+  def updateWith(key: K)(remappingFunc: Option[V] => Option[V]): Unit = {
+    val computeFunction = new java.util.function.BiFunction[K, V, V] {
+      override def apply(k: K, v: V): V = {
+        trieMap.get(key) match {
```
Comment (filipwiech) [inline mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala:21]: Not sure, but maybe if inside this function we referenced `k` instead of `key` (so input of `computeFunction` rather than the `updateWith` method) we could then extract it into a constant defined on the `AtomicTrieMap` class level so that there's no need to allocate a new one each time an `updateWith` is called? :thinking:

Comment (jkciesluk) [reply mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala:21]: I don't see its possible, because we need to pass it `remappingFunc`.
We could change `trieMap` type to `TrieMap[K, Set[V]]`, since we only use it like this and add a `append` function, that would look the way you suggested 🤔

Comment (filipwiech) [reply mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala:21]: Sorry, you're absolutely right, somehow I missed the `remappingFunc`. :+1:

File: mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+              case None =>
+                trieMap.remove(key)
+            }
+          case None =>
+            remappingFunc(None).foreach(trieMap.update(key, _))
+        }
+        null.asInstanceOf[V]
+      }
+    }
+    concurrentMap.compute(key, computeFunction)
```
Comment (tgodzik) [inline mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala:35]: Doesn't this duplicate the amount of data we hold? Should we just use ConcurrentMap instead?

Comment (jkciesluk) [reply mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala:35]: Size of concurrentMap is always 0, its only used to provide atomic updateWith for trieMap.
I don't think we should simply replace TrieMap with ConcurrentMap, because `get` on TrieMap should be faster

**Review by tgodzik — APPROVED** [review]
Thanks for explaning!

File: mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala (hunk 2 above)
Comment (kpodsiad) [reply mtags/src/main/scala/scala/meta/internal/mtags/AtomicTrieMap.scala:35]: > Size of concurrentMap is always 0, its only used to provide atomic updateWith for trieMap.

Can you ELI5 this? :D 

> get on TrieMap should be faster

Can you share some source about that, benchmark maybe? I did some reading, found https://medium.com/@igabaydulin/java-concurrenthashmap-vs-scala-concurrent-map-e185e8a0b798 which then led me to https://github.com/scala/scala/pull/10027 and I need to read everything once again because I'm already lost. Best part - those are not even about TrieMap 😅
